File: Python3/_316_RemoveDuplicateLetters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        lastIndex = {}
        seen = {}
        for i in range(len(s)):
            lastIndex[s[i]] = i
        
        stk = []
        for i in range(len(s)):
            if s[i] in seen and seen[s[i]] is True:
                continue
            while len(stk) >0 and stk[-1] > s[i] and lastIndex[stk[-1]] > i:
                seen[stk.pop()] = False
            stk.append(s[i])
            seen[s[i]] = True

        res = ""
        while len(stk) > 0:
            res = stk.pop() + res
        return res

test = Solution()
print(test.removeDuplicateLetters("cbacdcbc"))
print(test.removeDuplicateLetters("bcabc"))

File: Javascript/_316_RemoveDuplicateLetters.js
/**
 * @param {string} s
 * @return {string}
 */
 var removeDuplicateLetters = function(s) {
    lastIndex = {};
    seen = {};
    for(let i = 0; i<s.length; i++){
        lastIndex[s.charAt(i)] = i;
    }
    stk = [];
    
    for(let i = 0; i<s.length; i++){
        if(seen[s.charAt(i)]) continue;
        while(stk.length>0 && stk[stk.length-1] > s.charAt(i) && lastIndex[stk[stk.length-1] ] > i){
            seen[stk[stk.length-1]] = false;
            stk.pop();
        }
        stk.push(s.charAt(i));
        seen[s.charAt(i)] = true;
    }

    res = "";
    for(let i = 0; i<stk.length; i++){
        res += stk[i];
    }

    return res;
};


console.log(removeDuplicateLetters("cbacdcbc"));
console.log(removeDuplicateLetters("bcabc"));